Added a mechanism to check linearity of equations passed to linear_eq_to_matrix

czgdp1807 opened the pull request:
fixes #15195 
<!-- BEGIN RELEASE NOTES -->

* solvers
  * `linear_eq_to_matrix` checks if equations are actually linear, raising an error if they are not, instead of silently ignoring nonlinear terms

<!-- END RELEASE NOTES -->

gdp1807 committed:
Added a mechanism to check linearity in linear_eq_to_matrix

added a mechanism to check if the equations passed
to linear_eq_to_matrix method are linear or not. If they
are not linear then appropriate error messages are raised.

asmeurer commented:
It looks like there was a bug in the bot. I have fixed it at https://github.com/sympy/sympy-bot/pull/19. In the meantime, please do not use the first person in the release notes ("I have fixed"). See also 
https://github.com/sympy/sympy/wiki/Writing-Release-Notes

asmeurer commented:
Also:

- State in the pull request title what is fixed in the pull request. Do not put issue numbers in the pull request title
- To automatically close an issue, you need to write "fixes #14987" exactly. Writing the word "issue" between "fixes" and the number breaks GitHub's logic and prevents it from automatically being closed when the pull request is merged. 
- You can push to the same pull request. Do not open a new pull request for the same change. 
- Write tests for the change as well.

Abdullahjavednesar commented on sympy/solvers/solveset.py:
```diff
@@ -2805,4 +2827,4 @@ def nonlinsolve(system, *symbols):
         result = substitution(
             polys_expr + nonpolys, symbols, exclude=denominators)
-        return result
+        return result
```
Add a new line at the end of file.

czgdp1807 commented:
Sure I will add a new line at the end of the file.

czgdp1807 committed:
Added a new line at the end of the file.

Yathartha22 commented:
- "fixes #14987" should be a part of the description. Also add a brief description of what your changes are and how it solves the issue.
- The title should provide the aim of the PR not the issue number
- Release notes should not contain something like "I". Use something like:

```
* solvers
  * added check for linearity of equations before evaluating in `linear_eq_to_matrix()`
```
- Add tests.

czgdp1807 commented:
I think I have to work a little bit more on this PR. Correct test cases are misclassified as non-linear equations. I will edit and make all the suggested changes here only.

czgdp1807 committed:
attempt to debug linear_eq_to_matrix

made an attempt to debug linear_eq_to_matrix so that
 linear equations are not misclassified as non-linear equations

czgdp1807 committed:
made checking linearity optional

made the checking linearity optional by adding an optional argument **options

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -1705,6 +1707,14 @@ def linear_eq_to_matrix(equations, *symbols):
         symbols = symbols[0]
 
+    # setting default value for check_linearity
+    if 'check_linearity' not in options:
```
`check_linearity = options.get('check_linearity', False)` is the way to set the flag without modifying the passed dictionary.

czgdp1807 replied:
I have implemented this and made the changes.

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -1705,6 +1707,14 @@ def linear_eq_to_matrix(equations, *symbols):
+
+    if options['check_linearity']:
+        # removing duplicates from symbols
+        symbols = list(set(symbols))
```
Seems like this should always be done but you should retain the order of the passed symbols. `symbols = uniq(symbols)` would be preferred (with `uniq` coming from sympy.utilities.iterables).

czgdp1807 replied:
I have implemented it in a different way. Please check and give your suggestions for improvements.

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -1723,6 +1733,27 @@ def linear_eq_to_matrix(equations, *symbols):
         # append constant term (term free from symbols)
         coeff_list.append(-f.as_coeff_add(*symbols)[0])
 
+        if options['check_linearity']:
```
This reconstruction is not robust enough. Consider `eq = x + x*(y + 1)` which is linear in x but has a coefficient of x of `y + 2`; the expression that this code would test is `eq - x*(y + 2)` but that is not zero. There are some helpers like `degree` or `solve_linear` that are more careful about such equations.

czgdp1807 replied:
I have implemented solve_linear(). Now it works correctly for the equation said. However, there is a bug in coeff method. `(x + x*(y + 1)).coeff(y)` returns 0, but it should be x. Can I create an issue for this?

smichr replied:
> issue

No. coeff is fast and dumb by design, as stated in docstring "The matching is exact:" Your option is to first use expand on the equations -- that is the only way to know what the coefficients of the symbols are. We have not yet implemented a method of giving the exact order when there are power of monomials as in `x + (x + 1)**3`. This is perhaps why expansion is used by Poly at instantiation. If someone is working with a linear system, then the expansion step should be quick and assure you have accurate results.

czgdp1807 committed:
made the linear_eq_to_matrix more robust

made the linear_eq_to_matrix more robust
by using solve_linear() method

czgdp1807 commented:
@Yathartha22 I tried your method and found the following:-
1 . Equations like `x + sin(y)` will raise `PolynomialError: sin(y) contains an element of the generators set`.
2. Equations like `x*y` will be misclassified as linear even though its non-linear in symbols = [x , y].

czgdp1807 commented:
But @cbm755 said that I should avoid the use of derivatives in my code for checking linearity. What should I do to proceed?

cbm755 commented:
>  ... said I should avoid the use of derivatives

It just seems expensive.  But if that's the only way to do it...

Note, there are various changes [requested above](https://github.com/sympy/sympy/pull/14999#issuecomment-409159262) about this PR.

jksuom commented:
How about the following strategy: 

- Compute the list of coefficients as it is currently done (`f.coeff(symbol)`).
- Form the linear combination of symbols using the listed coefficients and subtract that from the original.
- If the difference is independent of the symbols, return it together with the coefficients, otherwise raise an exception.

czgdp1807 commented:
@jksuom I have implemented the same strategy as you are suggesting in this PR.

czgdp1807 commented:
@cbm755 I will replace options dict by skip_check shortly. Thanks for your suggestions. I used options so that more choices can be included in future.

cbm755 commented:
I think we need `skip_check` **only if** he additional checking is significantly more expensive.  Otherwise, it should not be added (IMHO).

czgdp1807 commented:
Can anyone tell me which one is better in efficiency and speed:
1. First expanding the equation using pre-defined method and then finding the coefficients using coeff method.
2. Finding the coefficients using derivatives.
I will implement the one which is more efficient.

czgdp1807 committed:
Optimized linear_eq_to_matrix 

Derivatives have been implemented in linea_eq_to_matrix to check_linearity with accuracy

czgdp1807 committed:
Added regression test for issue #14987

czgdp1807 committed:
Update test_solveset.py

Yathartha22 commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -1587,6 +1587,13 @@ def test_issue_10397():
+def test_issue_14987():
+    try:
+        linear_eq_to_matrix([x**2], [x], check_linearity = True)
+    except ValueError as e:
```
rather use: `raises(ValueError, lambda: linear_eq_to_matrix(......))`

Yathartha22 commented on sympy/solvers/solveset.py:
```diff
@@ -1705,6 +1707,16 @@ def linear_eq_to_matrix(equations, *symbols):
         symbols = symbols[0]
 
+    # setting default value for check_linearity
+    check_linearity = options.get('check_linearity', False)
```
I am not too convinced about using this `flag`. IMO every equation should be checked for being linear or not and not when this flag is set to `True`. Is there any specific reason for using it that I might be missing?

smichr replied:
Since this is a high-level routine (not called by anything else) I am in favor of a default check that assumptions are correct.

czgdp1807 commented:
I have used the that flag to tackle performance issues. The performance would be degraded for the code in SymPy that uses linear_eq_to_matrix. I will check that how much the performance is affected if we set check _linearity to true from the original code.

czgdp1807 commented:
I ran the code for 100,000 times for a non-linear equation and the time it took to execute `linear_eq_to_matrix` when `check_linearity = False`, was 55.197s and when `check_linearity=True`, was 222.77s. My personal view would be to keep the flag optional because there can be methods inside sympy code base that can slow down is we remove the flag.

czgdp1807 committed:
Updated test_issue_14987

czgdp1807 commented:
@asmeurer SymPy bot was not giving earlier but when with my latest commit it is failing some checks. What is the reason behind it?

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -1705,6 +1708,16 @@ def linear_eq_to_matrix(equations, *symbols):
+
+    if check_linearity:
+        # removing duplicates from symbols
+        symbs = uniq(symbols)
```
```
symbs = list(uniq(symbols))
```

czgdp1807 replied:
I will implement this.

czgdp1807 committed:
made checking linearity default in linear_eq_to_matrix

czgdp1807 committed:
removed check_linearity from test_issue_14987

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -1718,10 +1723,30 @@ def linear_eq_to_matrix(equations, *symbols):
         coeff_list = []
         for symbol in symbols:
-            coeff_list.append(f.coeff(symbol))
+            coeff_list.append(f.diff(symbol))
```
You can check the linearity now by seeing `if check_linearity and coeff_list[-1].free_symbols & set(symbols): raise Error`.

If this passes, you have a linear coefficient of an equation which was linear or had a removable singularity (e.g. f(x) = x*(1/x + 1)` has a singularity at f(0)`. You can make future derivatives faster by replacing symbol with 0. If you get S.NaN there was a removable singularity and if you want to ignore it, don't do the replacement.

```
newf = f.xreplace({symbol: 0})
if newf is not S.NaN:
    f = newf
```

smichr replied:
...or better, expand this portion of the expression to get rid of the singularity:

```
newf = f.xreplace({symbol: 0})
if newf is S.NaN:
    i, d = f.as_independent(symbol)
    f = f.func(i, expand_mul(d)).xreplace({symbol: 0})
```

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -1718,10 +1723,30 @@ def linear_eq_to_matrix(equations, *symbols):
 
         # append constant term (term free from symbols)
         coeff_list.append(-f.as_coeff_add(*symbols)[0])
+        
```
if you follow the suggestions above, this part will not be necessary

smichr replied:
(i.e. the code to check the coefficients with `solve_linear`

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -1718,10 +1723,30 @@ def linear_eq_to_matrix(equations, *symbols):
+            coeff_list.append(f.diff(symbol))
 
         # append constant term (term free from symbols)
         coeff_list.append(-f.as_coeff_add(*symbols)[0])
```
If you follow the suggestions above, you will just append `-f` because all the terms involving any symbol of interest have been removed.

smichr reviewed (commented):
suggestions to do checking as the coefficients are computed

czgdp1807 commented:
I will implement your suggestions shortly. Thanks for your feedback @smichr . :-)

czgdp1807 committed:
made changes in linearity checking of linear_eq_to_matrix

czgdp1807 committed:
removed import of solve_linear

czgdp1807 commented:
@smichr I have implemented the changes you suggested. Please give suggestions for further improvement.

Abdullahjavednesar commented:
@czgdp1807 can you please resolve conflicts?

czgdp1807 committed:
attempt to resolve conflicts

czgdp1807 commented:
I have made an attempt. Please give your review.

smichr committed:
update f as process continues

At the end of the loop through symbols, f will (if updated properly) contain
only the constant term. A test of an equation like `x*(-3/x + 1) + 2*y - a` for symbols 
x and y would be good to add.

czgdp1807 committed:
attempt to remove build errors

czgdp1807 committed:
more robust method to remove build errors

czgdp1807 committed:
updated test_issue_14987 with an extra test

czgdp1807 commented:
@smichr I have added the required test as well as made an attempt to remove the build errors which appeared after the commit `4632cdc`.

smichr commented:
`f=sympify(newf)` should not be needed; that it *is* needed might mean that the xreplace is returning an int instaed of Integer. At your interactive session try `type(x.xreplace({x:0}))`. Do you get `int` or `Integer`. If `int`, then the fix is to replace the 0 in the first xreplace line with `S.Zero`; in the second case you could also use it but it shouldn't be needed since the `f.func` will handle the conversion...but it wouldn't hurt. I'll just try those mods here.

czgdp1807 commented:
@smichr I got your point. Making suitable changes.

smichr committed:
change 0 to S.Zero in xreplace

czgdp1807 commented:
Thanks @smichr :-) I was just going to make the same commit.

smichr commented:
So how do the times compare now? Are you still thinking of adding a flag to disable checking?

An interesting comparison  would be with this routine and the following:

```
def linjac(eqs, syms):
    E = Matrix(eqs)
    j = E.jacobian(syms)
    assert not j.free_symbols & set(syms)
    c = [-i.xreplace(dict(zip(syms, [S.Zero]*len(syms)))) for i in E]
```

It *seems* like making `f` simpler as you go is better (whereas the jacobian is not going to do that).

smichr commented:
Think more about what @jksuom suggested. There is wisdom in taking the fast simple path and only taking the more robust (slower) path when necessary...and also in remembering that the slow step is going to be in solving these equations. Building the equations is likely a minor part of the whole process.

Here are some routines to consider:

```python
def co(eq, syms):
    # all equations are flat wrt symbols
    # 2*x + y
    rv = []
    free =  set(syms)
    for e in eqs:
        rv.append([])
        L = rv[-1]
        for s in syms:
            L.append(e.coeff(s))
        L.append(e - Add(*[c*s for c, s in zip(L, syms)]))
        assert not L[-1].free_symbols & free
    return rv

def coe(eq, syms):
    # all equations are linear wrt symbols but not flat
    # (x + 1)*y is linear in x
    return co([expand_mul(e) for e in eqs], syms)

def coed(eq, syms):
    # all equations are linear wrt symbols but not flat
    # and might have removable singularities: x*(1/x+1) + y
    for e in eqs:
        assert not denoms(e, *syms)
    return coe(eq, syms)
```

czgdp1807 commented:
@smichr I earlier implemented the method suggested by @jksuom. Then I further made changes in my PR.

cbm755 commented:
If you think @jksuom's approach won't work, please provide an actual concrete example (which would become a test case).

czgdp1807 commented:
Thanks again @smichr. I will look more into my PR and maje suitable changes as soon as possible. :-)

cbm755 commented:
Why should this code care about removable singularities?  Seems like Someone Else's Problem (the caller can `simplify` if it wants to clean those up).

smichr commented:
OK...keep it really simple:

```
free = set(symbols)
zero = dict(zip(symbols, [S.Zero]*len(symbols)))
if len(symbols) != len(free):
    raise ValueError('symbols must be unique')
for symbol in symbols:
    coeff_list.append(f.coeff(symbol))
    common = coeff_list[-1].free_symbols & free
    if common:
        fex = linear_eq_to_matrix(f.expand())
        try:
            fex = linear_eq_to_matrix(f.expand(), symbols)
            raise ValueError('use expand on the equation(s) before calling this routine')
        except ValueError:
            raise ValueError('equation(s) nonlinearity detected in {%s}' % common)
...
coeff_list.append(-f.xreplace(zero))
```

czgdp1807 commented:
@smichr Is this patch in your comment the final one?

smichr commented:
> the final one

Please forgive the derivative-on-the-brain. If you looked at #15049 you would understand why this might be ;-)

cbm755 commented:
> Is something linear if it has a nonlinear term that does not cancel by addition but cancels by division?

No.  That would be nonlinear, and the new expression (after cancelling) is linear.  The solutions in the two cases may be different.

Perhaps we should add doctests to the docstring that make this explicit:
```
If the equations are nonlinear, a ValueError will be raised:

>>> linear_eq_to_matrix(x**2 + x + 3, x)
ValueError ...

This routine does not generally simplify expressions, for example, consider
the nonlinear equations

>>> eqns = [Eq((x**2-3*x)/(x-3) , 3), Eq(y**2 - 3*y - y*(y - 4) + x, 4)]
>>> linear_eq_to_matrix(eqns, [x, y])
ValueError ...

However, if simplifying these equations will discard the removable singularity
in the first and reveal the linear structure of the second
>>> eqns2 = [simplify(eqn) for eqn in eqns]
>>> eqns2
[x = 3, x + y = 4]

>>> linear_eq_to_matrix(eqns2, [x, y])
⎛⎡1  0⎤  ⎡3⎤⎞
⎜⎢    ⎥, ⎢ ⎥⎟
⎝⎣1  1⎦  ⎣4⎦⎠
```

cbm755 commented:
Also, just FYI, not for the docs:
```py
>>> solve(eqns[0],x)
[]

>>> solve(eqns2[0],x)
[3]
```
These are correct answers and should make it clear the difference between a linear equation and an equation with a removable singularity that happens to be linear upon removal.

smichr commented:
🔟 Nice example @cbm755 . This equation is also one that expansion alone will not identify as having a removable singularity: `cancel` is necessary.

cbm755 commented:
So can we agree *this code* should stop worrying about such things? 

`linear_eq_to_matrix` It is a *structural* tool: if the equation as-is is linear, it will give the matrix form.  Otherwise it will raise a ValueError.

All of your examples in the previous comment should be test cases: in each case it should raise ValueError.

cbm755 commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -1587,6 +1587,11 @@ def test_issue_10397():
 
+def test_issue_14987():
+    raises(ValueError, lambda: linear_eq_to_matrix([x**2], x))
+    assert linear_eq_to_matrix([x*(-3/x + 1) + 2*y - a], [x, y]) == (Matrix([[1, 2]]), Matrix([[a + 3]]))
```
this test is incorrect, it raise ValueError

cbm755 commented on sympy/solvers/solveset.py:
```diff
@@ -1705,6 +1706,8 @@ def linear_eq_to_matrix(equations, *symbols):
     if hasattr(symbols[0], '__iter__'):
         symbols = symbols[0]
 
+    symbols = list(uniq(symbols))
```
Not sure `uniq` is safe to do to user input: the order must not be changed for example.

See @smichr has some code which checks the length and raises an exception: I think that is the right thing.

czgdp1807 committed:
Updated linear_eq_to_matrix

czgdp1807 committed:
Updated test_issue_14987

czgdp1807 committed:
Added tests in test_issue_14987

czgdp1807 commented:
I have made the changes as suggested in the comments above. Please have a look at them.

smichr commented:
significant error in not detecting `x**2` as invalid. How about

```python
def linco(f, *symbols):
    coeff_list = []
    const, d = f.as_independent(*symbols, as_Add=True)
    for s in symbols:
        d, spart = d.as_independent(s, as_Add=True)
        if not spart:
            coeff_list.append(spart)
            continue
        try:
            c, (x,) = spart.as_coeff_mul(s)
            assert x == s
            coeff_list.append(c)
        except (ValueError, AssertionError):
             raise ValueError('nonlinear; simplification might help')
    coeff_list.append(-const)
    return coeff_list
```

smichr committed:
make coefficient detection more robust

Although linsolve expands expressions via Poly and detects non-linearity
before passing to linear_eq_to_matrix, the latter does its own non-linearity
checking without using expansion so it will detect removable singularities
and systems that have not been simplified.

smichr committed:
remove unused imports

smichr commented:
I see that this routine is called only by linsolve which already does the non-linearity checking. So this routine now does so, too, but does so in a more conservative fashion by not using any simplification of the input.

smichr committed:
Update test_solveset.py

smichr committed:
Update test_solveset.py

smichr commented:
I will commit in 24 h if there is no more feedback.

cbm755 commented:
hold off please, I'd like to review this properly, especially as its more than the three or four line fix I expected.

smichr committed:
cross-term nonlinearity now detected in linsolve

Expansion is retained in linsolve and error detection is now relegated to liner_eq_to_matrix.

smichr committed:
Merge branch 'master' into patch-19

smichr commented:
> more than the three or four line fix

* linsolve will now detect cross-term non-linearity (Poly(x*y,x,y).degree() == 1 so this was not caught. One option would be to use `total_degree` the other is to just let it be detected (without the need for Poly) in `linear_eq_to_matrix`. The expansion that would have been done by Poly is now handled with `_mexpand` with a recursive=True option.

* extraction of coefficients and detection of non-linearity in `linear_eq_to_matrix` is handled without any more reconstruction of Add's that was originally done.

smichr committed:
update tests

smichr commented:
I sent a PR to this branch that updates tests and passes locally.

czgdp1807 committed:
Merge pull request #2 from smichr/patch-19

update tests

czgdp1807 commented:
@smichr I have merged the PR created into `czgdp1807:patch-19`. Please have a look.

smichr commented:
> I have merged the PR

Thanks. Let's wait for @cbm755 to look things over.

smichr commented:
ping @cbm755

smichr commented:
@asmeurer, I would like to see this get into 1.3. It fixes some bugs that would pass without comment from the existing routines.

smichr commented:
@Yathartha22 , any chance you could give this a look over?

Yathartha22 commented:
Sure @smichr I will have a look at it and update you soon.

cbm755 commented on sympy/solvers/solveset.py:
```diff
@@ -1829,84 +1832,128 @@ def linear_eq_to_matrix(equations, *symbols):
 
-    for equation in equations:
+    for row_no, equation in enumerate(equations):
         f = sympify(equation)
```
Docs explicity say simplifying not done...  Maybe should say some milder?  "This routine might not fully simplify expressions"

cbm755 reviewed (commented):
Here are some comments from proofreading; have not looked at correctness.

cbm755 commented on sympy/solvers/solveset.py:
```diff
@@ -1829,84 +1832,128 @@ def linear_eq_to_matrix(equations, *symbols):
+    if has_dups(symbols):
+        raise ValueError(filldedent('''
+            Symbols must be unique.
+            '''))
```
Is there a test for this?

smichr replied:
added

cbm755 commented:
Ok, so I've looked over this and it looks reasonable.  Why is it better than what @jksuom and I suggested?
```diff
diff --git a/sympy/solvers/solveset.py b/sympy/solvers/solveset.py
index 3cdb14775..98f156e48 100644
--- a/sympy/solvers/solveset.py
+++ b/sympy/solvers/solveset.py
@@ -1909,43 +1909,20 @@ def linear_eq_to_matrix(equations, *symbols):
         if isinstance(f, Equality):
             f = f.lhs - f.rhs
 
-        # initialize args for each coeff to empty
-        coeff_list = [[] for s in symbols]
-
-        # get pointers for locations of symbols
-        ix = dict(zip(symbols, range(len(symbols))))
-
-        # separate into terms
-        # e.g. (x + 2*y + x*y + 4*x**3 + 6).as_coeff_add(x,y)
-        # (6, (x, 2*y, 4*x**3, x*y))
-        const, terms = f.as_coeff_add(*symbols)
-
-        # store constant term
-        coeff_list.append([-const])  # 6 as -6
-
-        # collect coefficients of symbols
-        valid_form = False
-        for t in terms:
-            c, factors = t.as_coeff_mul(*symbols)
-            if len(factors) != 1:
-                # e.g. x*y -> (1, (x, y))
-                break
-            if factors[0] not in symbols:
-                # e.g. x**3 -> (1, (x**3,))
-                break
-            coeff_list[ix[factors[0]]].append(c)
-        else:
-            valid_form = True
-
-        # check if all equations were in valid form
-        if not valid_form:
-            __ = (t, set(symbols))
-            raise ValueError(filldedent('''
-                The term %s is nonlinear in %s
-                ''' % __))
-
-        # rebuild coefficients
-        coeff_list = [Add(*i) for i in coeff_list]
+        coeff_list = []
+        for symbol in symbols:
+            c = f.coeff(symbol)
+            if any(c.has(s) for s in symbols):
+                    raise ValueError('The expression %s seems to be nonlinear in %s' %
+                                     (f, symbol))
+            coeff_list.append(c)
+
+        const = simplify(f - Add(*[c*s for (c,s) in zip(coeff_list, symbols)]))
+        if any(const.has(s) for s in symbols):
+                raise ValueError('The expression %s seems to be nonlinear in %s' %
+                                 (f, s))
+        coeff_list.append(-const)
 
         # insert coefficients as a row in M
         M = M.row_insert(row_no + 1, Matrix([coeff_list]))
```

Against master, this is the "three or four line fix":
```diff
@@ -1895,18 +1912,25 @@ def linear_eq_to_matrix(equations, *symbols):
         # Extract coeff of symbols
         coeff_list = []
         for symbol in symbols:
-            coeff_list.append(f.coeff(symbol))
+            c = f.coeff(symbol)
+            if any(c.has(s) for s in symbols):
+                    raise ValueError('The expression %s seems to be nonlinear in %s' %
+                                     (f, symbol))
+            coeff_list.append(c)
 
-        # append constant term (term free from symbols)
-        coeff_list.append(-f.as_coeff_add(*symbols)[0])
+        const = simplify(f - Add(*[c*s for (c,s) in zip(coeff_list, symbols)]))
+        if any(const.has(s) for s in symbols):
+                raise ValueError('The expression %s seems to be nonlinear in %s' %
+                                 (f, s))
+        coeff_list.append(-const)
 
        # insert equations coeff's into rows
        M = M.row_insert(row_no, Matrix([coeff_list]))
        row_no += 1
 
     # delete the initialized (Ist) trivial row
     M.row_del(0)
     A, b = M[:, :-1], M[:, -1:]
```

1.  This seems to pass the new tests.  So if there are examples for which this fails and your's works, we should have those as tests too!

2.  Is there an efficiency difference?  Maybe your approach has quadratic runtime and mine is cubic?

I really don't have a problem with the new implementation; would just like to see concrete tests for what is wrong with a minor change instead of a rewrite.

cbm755 commented on sympy/solvers/solveset.py:
```diff
@@ -1829,84 +1832,128 @@ def linear_eq_to_matrix(equations, *symbols):
 
-    M = Matrix([symbols])
+    if not hasattr(equations, '__iter__'):
+        equations = [equations]
```
This breaks the following test which works on master.
```py
>>> a,b,c = symbols('a,b,c')
>>> equations = Matrix([[a*x + b*y - 7], [5*x + 6*y - c]])
>>> linear_eq_to_matrix(equations, [x, y])
⎛⎡a  b⎤  ⎡7⎤⎞
⎜⎢    ⎥, ⎢ ⎥⎟
⎝⎣5  6⎦  ⎣c⎦⎠
```
That should be added as a test case.

Perhaps it should be `if not isinstance(equations, MatrixBase) and not hasattr(...)`?

cbm755 replied:
Also, I don't see a test for this change.  How about:
```py
assert linear_equation_to_matrix(3*x + 2*y - 5, [x, y]) = 
    (Matrix([[3, 2]]), Matrix([[5]]))
```

Yathartha22 commented on sympy/solvers/solveset.py:
```diff
@@ -1829,84 +1832,128 @@ def linear_eq_to_matrix(equations, *symbols):
+    >>> [e.simplify() for e in eqns]
+    [x - 3, x + y - 4]
+
+    Any needed simplification (expansion, factoring, etc...) needed
```
Could be : `Any simplification (expansion, factoring, etc...) needed..`

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -1829,84 +1832,128 @@ def linear_eq_to_matrix(equations, *symbols):
+    [    0]])
+
+    This routine does not simplify expressions. For example, consider
+    the nonlinear equations.
```
done

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -2053,6 +2100,16 @@ def linsolve(system, *symbols):
     EmptySet()
 
+    * Although removable singularities and nonlinearity
+      will not raise an error, non-linear systems will:
```
modified

cbm755 replied:
Is this change in [the pr](https://github.com/czgdp1807/sympy/pull/3)?  I didn't see it.

cbm755 replied:
This sentence is still confusing.  I think the change you made didn't make it to the latest commits.

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -2082,19 +2139,7 @@ def linsolve(system, *symbols):
-                    ''') % eq)
+            system = [_mexpand(i, recursive=True) for i in system]
             system, symbols, swap = recast_to_symbols(system, symbols)
             A, b = linear_eq_to_matrix(system, symbols)
```
ok

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -1829,84 +1832,128 @@ def linear_eq_to_matrix(equations, *symbols):
+    if has_dups(symbols):
+        raise ValueError(filldedent('''
+            Symbols must be unique.
+            '''))
```
I write it like this so the user doesn't have to see the error message printed twice: once in the traceback and once in the message. Here is what it looks like now:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "\sympy\solvers\solveset.py", line 1902, in linear_eq_to_matrix

ValueError:
Symbols must be unique.
```

And here is what the one-liner looks like:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "\sympy\solvers\solveset.py", line 1907, in linear_eq_to_matrix
    raise ValueError(filldedent('''Symbols must be unique.'''))
ValueError:
Symbols must be unique.
```

cbm755 replied:
I meant: `raise ValueError('Symbols must be unique')`.  I'm skeptical this `ValueError(filldenent('''` stuff is really preferable, at least in that case.

smichr commented:
> would just like to see concrete tests

It's not the scaling that worries me as much as the cost of using simplify to do the job of testing for linearity. Here is a simple test:
```python
def do(n, m):
    from random import shuffle
    numbers = list(range(n))
    tot = 0
    for do in range(n):
        shuffle(numbers)
        eqs = []
        s = [Dummy() for i in range(m)]
        for i in range(m):
            const = numbers[m+1]
            #const = sin(const)
            #const = sin(Dummy() + const)
            eqs.append(Add(*[s[j]*numbers[j] for j in range(m)]+[const]))
        t = time() <-- start timer
        ans = linear_eq_to_matrix(eqs, s)
        tot += time()-t <-- end it
    return tot

for m in (3,6,12):
    do(1000,m)
```
The results of this are:

```
this PR
3.2
9.1
43.2

master (unmodified)
3.3
9.1
43.6

suggested mods
4.6
12.2
53.8
```

So the short mods are a little slower. But this is the best case scenario with nothing for simplify to do.
If the constant is made a little more complicated the cost of using simplify become apparent:

```
change constant to sin(constant)
4.7
13.6
56.6

change constant to sin(Dummy() + constant)
48
101.5
280
```
The suggestion of simplify is apropos here:

> If your algorithm relies on "simplification" (whatever it is), try to
    determine what you need exactly  -  is it powsimp()?, radsimp()?,
    together()?, logcombine()?, or something else? And use this particular
    function directly, because those are well defined and thus your algorithm
    will be robust.

We know a lot about what we are looking for and what we will reject. By avoiding the blanket simplification to detect nonlinearity, our times are nearly the same as for the simplest case considered first:
```
3.9
13.2
58.2
```

cbm755 commented:
> By avoiding the blanket simplification to detect nonlinearity,
> our times are nearly the same as for the simplest case

I see, and my approach will always be fragile without this blanket simplification.  Thank you for the detailed tests.

smichr commented:
For sake of brevity, however, there may be a better way. If we know we aren't going to simplify then a simple expand_mul, after testing for denoms, will give equations that can be taken apart without needing to be simplified to test that the coefficient has been found.

asmeurer commented:
@smichr did you try to push changes up to @czgdp1807's branch. It looks like sympy/sympy now has a patch-19 branch. I would recommend setting your `origin` to read-only to avoid accidentally pushing up changes.

smichr commented:
> did you try to push changes up

Locally, I downloaded his repo, created a new branch of his patch-19, edited it, pushed it via git, opened a new pull request against czgdp1807's (copied and pasted that into the URL) and created the request: https://github.com/czgdp1807/sympy/compare/patch-19...smichr:patch-19?expand=1. It created the PR that is [here](https://github.com/czgdp1807/sympy/pull/3).

Then I told him about it. I didn't commit anything. 

Perhaps @cbm755 knows something about this?

Please let me know if there is evidence that I did something wrong, otherwise I won't know what to do to correct it.

smichr commented:
@asmeurer , can that patch-19 branch just be deleted?

asmeurer commented:
I think you somehow pushed it to sympy/sympy in the process on accident. I can't say exactly how it happened without seeing your command history. That's why I recommend setting your sympy/sympy remote to read-only (git://) so it can't happen accidentally. 

The branch can be deleted. I didn't want to do it because it looked like it has commits that aren't here, so I didn't want to lose them.

cbm755 commented:
@smichr if I did it, it was by accident (and I think my sympy/sympy remote is readonly), no problem from me if you delete it.

smichr commented:
Where is the info after executing "git remote set-url origin git://github.com/sympy/sympy.git" stored?

When I try to execute it I get:
```
$ git remote set-url origin git://github.com/sympy/sympy.git
fatal: No such remote 'origin'
```

I checked my history and the commands that *recently* pushed patch-19 did not push anything to master...but I also checked my remotes and see `.git/logs/refs/remotes/origin/patch-19` and `, `.git/logs/refs/remotes/smichr/patch-19`, and `.git/logs/refs/remotes/czgdp1807/patch-19`. Seeing that "origin" one looks like the smoking gun to me. Apologies for that. It must have happened from the first patch that I made. Looking now at the branch on origin (which I have deleted) I see that it was updated "15 days ago" which would make sense: my most recent activity was only a few days ago.

smichr commented:
I *am* able to do this:
```
git remote add origin git://github.com/sympy/sympy.git
```

smichr commented:
This fixes a WrongResult. It should go in this release, if possible. We are waiting for @czgdp1807 to commit the PR that I sent. I will open a cleaned up PR this evening if nothing happens here.

smichr committed:
edits in response to review (1/2)

smichr committed:
edits in response to review (2/2)

smichr commented:
I just made the changes directly here and will commit once tests pass.

smichr commented:
@asmeurer , is it ok to have this based on 1.3?

asmeurer commented:
Unless it's a critical fix I would keep it in master. If it is critical, you can change the base of the PR.

smichr committed:
Update solveset.py

cbm755 commented:
This is not a regression...  And the errors have been there for many sympy versions.  So although I care a lot about fixing this, I would say its not critical for 1.3.

And there are still some wordings and strange (to me) formatting (all the filldedent stuff).

smichr committed:
Update test_solveset.py

smichr commented:
I cleaned everything up and made a new PR at #15212. If it's ok with you, @czgdp1807 , could this be closed?
